tamr_toolbox/low_latency/llm.py

```diff
@@ -0,0 +1,215 @@
+import json
+import logging
+import time
+from collections import defaultdict
+from typing import Dict, List, Optional, Union
+
+from tamr_unify_client import Client
+from tamr_unify_client.operation import Operation
+from tamr_toolbox.utils.operation import from_resource_id
+
+from tamr_toolbox.models.data_type import JsonDict
+
+LOGGER = logging.getLogger(__name__)
+
+
+def update_llm_data(
+    client: Client,
+    *,
+    project_name: str,
+    do_update_clusters: bool = True,
+    do_use_manual_clustering: bool = False,
+    **options,
+) -> Operation:
+    """
+    Updates data for LLM query if needed, based on latest published clusters.
+
+    Args:
+        client: Tamr client object (connected to primary Tamr port , i.e. 9100)
+        project_name: name of the project to be updated
+        do_update_clusters: whether to update clusters, default True
+        do_use_manual_clustering: whether to use externally managed clustering, default False
+        options: Options passed to underlying :class:`~tamr_unify_client.operation.Operation`
+    """
+
+    url = (
+        f"projects/{project_name}:updateLLM?updateClusters={do_update_clusters}"
+        f"&useManualClustering={do_use_manual_clustering}"
+    )
+    response = client.post(url)
+    if not response.ok:
+        message = (
+            f"LLM update for {project_name} failed at submission time: "
+            + response.json()["message"]
+        )
+        LOGGER.error(message)
+        raise RuntimeError(message)
+    operation_id = response.content.decode("latin1")
+    operation = from_resource_id(client, job_id=operation_id)
+
+    return operation.apply_options(**options)
+
+
+def poll_llm_status(
+    match_client: Client, *, project_name: str, num_tries: int = 10, wait_sec: int = 1
+) -> bool:
+    """
+    Check if LLM is queryable. Try up to num_tries times at 1s (or user-specified) interval.
+
+    Args:
+        match_client: a Tamr client set to use the port of the Match API
+        project_name: name of target mastering project
+        num_tries: max number of times to poll endpoint, default 10
+        wait_sec: number of seconds to wait between tries, default 1
+
+    Returns:
+        bool indicating whether project is queryable
+    """
+
+    url = f"/api/v1/projects/{project_name}:isQueryable"
+    counter = 0
+
+    # Poll endpoint to until project is queryable or num_tries reached
+    while counter < num_tries:
+
+        response = match_client.get(url)
+        queryable = response.content == b"true"
+
+        if queryable:
+            break
+
+        counter += 1
+        time.sleep(wait_sec)  # call api at wait_sec interval if project isn't yet queryable
+
+    return queryable
+
+
+def llm_query(
+    match_client: Client,
+    *,
+    project_name: str,
+    records: Union[JsonDict, List[JsonDict]],
```

Comment:
```suggestion
    records: List[JsonDict],
```
I think it's better here that we require a list of records be passed even if there is only a single one.  Allowing multiple types to be passed here can make type-inspection with IDE tools and any automatic type-enforcement less robust.  It also means the code can be slightly less complex.